File: apple/BoundlessRider/Sources/RiderShared/Onboarding/OnboardingRouter.swift
import BoundlessKit
import SwiftUI

/// Maps the current `OnboardingState` to the screen the Rider sees, wiring each affordance to the
/// view model. The router owns only *view* state (the field text and the two in-step
/// acknowledgements); the authoritative `OnboardingState` lives in the view model and is advanced
/// solely by the core (P4). Completion is **silent** — `.complete` routes to the primary surface
/// with no "all set" screen (voice-and-tone).
public struct OnboardingRouter: View {
    @State private var viewModel: OnboardingViewModel
    @State private var phone = ""
    @State private var code = ""
    // In-step view acknowledgements (the core state is unchanged until their Continue is tapped).
    @State private var permissionDeclinedAck = false
    @State private var autoUpdateEnabledAck = false

    public init(viewModel: OnboardingViewModel) {
        _viewModel = State(initialValue: viewModel)
    }

    public var body: some View {
        Group {
            if viewModel.state == .complete {
                PrimarySurfacePlaceholderView()
            } else {
                OnboardingScreenView(currentModel)
            }
        }
    }

    private var adminName: String? { viewModel.adminName }

    private var currentModel: OnboardingScreenModel {
        switch viewModel.state {
        case .freshInstall:
            return RiderOnboardingScreens.helperIntro(onContinue: viewModel.begin)

        case .phoneEntry:
            return RiderOnboardingScreens.phoneEntry(text: $phone) {
                Task { await viewModel.submitPhone(phone) }
            }

        case .phoneNotOnFile:
            return RiderOnboardingScreens.phoneNotOnFile(text: $phone, adminName: adminName) {
                Task { await viewModel.submitPhone(phone) }
            }

        case .deviceBinding:
            return RiderOnboardingScreens.deviceBinding(text: $code, adminName: adminName) {
                Task { await viewModel.submitCode(code) }
            }

        case .bindingFailed:
            return RiderOnboardingScreens.bindingFailed(text: $code, adminName: adminName) {
                Task { await viewModel.submitCode(code) }
            }

        case .permissions:
            if permissionDeclinedAck {
                return RiderOnboardingScreens.permissionsDeclined(adminName: adminName) {
                    Task { await viewModel.decideNotifications(allow: false) }
                }
            }
            return RiderOnboardingScreens.permissions(
                onAllow: { Task { await viewModel.decideNotifications(allow: true) } },
                onDecline: { permissionDeclinedAck = true }
            )

        case .autoUpdateStep:
            if autoUpdateEnabledAck {
                return RiderOnboardingScreens.autoUpdateEnabled(onContinue: viewModel.confirmAutoUpdate)
            }
            return RiderOnboardingScreens.autoUpdateStep(onContinue: { autoUpdateEnabledAck = true })

        case .belowMinVersion, .needsReauthHelp:
            // Terminal calm screen — no CTA, no form (AC8 / AC15 / P10).
            return RiderOnboardingScreens.calmHelp(adminName: adminName)

        case .complete:
            // Unreachable: `.complete` is handled above by the placeholder. Kept for exhaustiveness.
            return RiderOnboardingScreens.calmHelp(adminName: adminName)
        }
    }
}

/// The hand-off after a **silent** onboarding completion. The real Rider primary surface ("You're
/// coming tonight") is a later spec; this neutral placeholder asserts the constitution's "no
/// celebration of plumbing" — there is deliberately no "all set" copy here.
public struct PrimarySurfacePlaceholderView: View {
    public init() {}
    public var body: some View {
        Color(.systemBackground)
            .ignoresSafeArea()
            .accessibilityHidden(true)
    }
}

File: android/rider/shared/src/main/kotlin/app/boundless/rider/onboarding/OnboardingRouter.kt
package app.boundless.rider.onboarding

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import app.boundless.rider.i18n.RiderStrings
import kotlinx.coroutines.launch
import uniffi.boundless_ffi_kotlin.OnboardingState

/**
 * Maps the current [OnboardingState] to the screen the Rider sees, wiring each affordance to the
 * view model. The router owns only *view* state (the field text and the two in-step
 * acknowledgements); the authoritative [OnboardingState] lives in the view model and is advanced
 * solely by the core (P4). Completion is **silent** — `COMPLETE` routes to the primary surface with
 * no "all set" screen (voice-and-tone). The Android twin of `RiderShared.OnboardingRouter`.
 *
 * The composition root (MainActivity, deferred to T13-shell) supplies the [strings] resolver
 * (`AndroidRiderStrings`) and constructs the [viewModel] with the real conformers.
 */
@Composable
fun OnboardingRouter(viewModel: OnboardingViewModel, strings: RiderStrings) {
    val screens = remember(strings) { RiderOnboardingScreens(strings) }
    val scope = rememberCoroutineScope()
    var phone by remember { mutableStateOf("") }
    var code by remember { mutableStateOf("") }
    // In-step view acknowledgements (the core state is unchanged until their Continue is tapped).
    var permissionDeclinedAck by remember { mutableStateOf(false) }
    var autoUpdateEnabledAck by remember { mutableStateOf(false) }

    val adminName = viewModel.adminName
    when (viewModel.state) {
        OnboardingState.COMPLETE ->
            // Silent completion — the hand-off placeholder, no "all set" celebration.
            PrimarySurfacePlaceholder()

        OnboardingState.FRESH_INSTALL ->
            OnboardingScreenView(screens.helperIntro(onContinue = viewModel::begin))

        OnboardingState.PHONE_ENTRY ->
            OnboardingScreenView(
                screens.phoneEntry { scope.launch { viewModel.submitPhone(phone) } },
                fieldValue = phone,
                onFieldValueChange = { phone = it },
            )

        OnboardingState.PHONE_NOT_ON_FILE ->
            OnboardingScreenView(
                screens.phoneNotOnFile(adminName) { scope.launch { viewModel.submitPhone(phone) } },
                fieldValue = phone,
                onFieldValueChange = { phone = it },
            )

        OnboardingState.DEVICE_BINDING ->
            OnboardingScreenView(
                screens.deviceBinding(adminName) { scope.launch { viewModel.submitCode(code) } },
                fieldValue = code,
                onFieldValueChange = { code = it },
            )

        OnboardingState.BINDING_FAILED ->
            OnboardingScreenView(
                screens.bindingFailed(adminName) { scope.launch { viewModel.submitCode(code) } },
                fieldValue = code,
                onFieldValueChange = { code = it },
            )

        OnboardingState.PERMISSIONS ->
            if (permissionDeclinedAck) {
                OnboardingScreenView(
                    screens.permissionsDeclined(adminName) {
                        scope.launch { viewModel.decideNotifications(allow = false) }
                    },
                )
            } else {
                OnboardingScreenView(
                    screens.permissions(
                        onAllow = { scope.launch { viewModel.decideNotifications(allow = true) } },
                        onDecline = { permissionDeclinedAck = true },
                    ),
                )
            }

        OnboardingState.AUTO_UPDATE_STEP ->
            if (autoUpdateEnabledAck) {
                OnboardingScreenView(screens.autoUpdateEnabled(onContinue = viewModel::confirmAutoUpdate))
            } else {
                OnboardingScreenView(screens.autoUpdateStep(onContinue = { autoUpdateEnabledAck = true }))
            }

        // Terminal calm screens — no CTA, no form (AC8 / AC15 / P10).
        OnboardingState.BELOW_MIN_VERSION, OnboardingState.NEEDS_REAUTH_HELP ->
            OnboardingScreenView(screens.calmHelp(adminName))
    }
}
